java-dynamic-sqs-listener-api/src/main/java/com/jashmore/sqs/broker/MessageBroker.java

```diff
@@ -0,0 +1,55 @@
+package com.jashmore.sqs.broker;
+
+import com.jashmore.sqs.processor.MessageProcessor;
+import com.jashmore.sqs.retriever.MessageRetriever;
+
+import java.util.concurrent.Future;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Broker used to co-ordinate the retrieval and processing of messages from a remote queue.
+ *
+ * <p>If you were to consider this library as similar to a pub-sub system, this would be considered the broker that transports messages from the publisher
+ * to the subscriber.  It will request for messages from the {@link MessageRetriever} and delegate the processing to a corresponding {@link MessageProcessor}.
+ *
+ * <p>As this is a non-blocking class that will start the message process in a background thread there is the possibility of usage of this class in multiple
+ * threads. Therefore, implementations of this class must be thread safe.
+ */
+@ThreadSafe
+public interface MessageBroker {
+    /**
+     * Start the brokerage of messages from the queue to the processor.
+     *
+     * <p>This must be non-blocking and start the process on a background thread.
+     *
+     * <p>If this broker has already been started then any subsequent calls to this method will do no action until {@link #stop(boolean)} has been
+     * called. Once {@link #stop(boolean)} has been called, any calls to this method should spin up a new thread and it should not be dependent on the stop
+     * finally finishing.
```

Comment:
Also, the fact that I can restart it even if there are existing threads hanging on finishing processing seems ok but may be helpful to tell caller or give caller the ability to check # of existing threads still not cleaned up. I think this is edge casey and probably not a requirement currently. Just letting my thoughts spew out here though.

Reply:
Yeah this is a good point. I did it this way for simplification really. Once you stop it we just assume that it will stop eventually and I am allowed to start it again if I want to and I shouldn't really need to care that an existing thread is shutting down. Or maybe I do? I have been on the fence for this for a while

Reply:
This is fine for now for simplicity. If it becomes an issue you can code for it then.